include/deal.II/matrix_free/cuda_matrix_free.templates.h

```diff
@@ -169,162 +152,137 @@ namespace CUDAWrappers
+                                   data->inv_jacobian[color]);
+
+      auto cell = graph.cbegin(), end_cell = graph.cend();
+      for (unsigned int cell_id = 0; cell != end_cell; ++cell, ++cell_id)
         {
-          data->q_points[color] = q_points;
-        }
+          (*cell)->get_dof_indices(local_dof_indices);
+          // When using MPI, we need to transform the local_dof_indices, which
+          // contains global dof indices, to get local (to the current MPI
+          // process) dof indices.
+          if (partitioner)
+            for (auto &index : local_dof_indices)
+              index = partitioner->global_to_local(index);
+
+          for (unsigned int i = 0; i < dofs_per_cell; ++i)
+            lexicographic_dof_indices[i] =
+              local_dof_indices[lexicographic_inv[i]];
+
+          const ArrayView<
+            dealii::internal::MatrixFreeFunctions::ConstraintKinds>
+            cell_id_view(constraint_mask_host[cell_id]);
+
+          hanging_nodes.setup_constraints(*cell,
+                                          partitioner,
+                                          {lexicographic_inv},
+                                          lexicographic_dof_indices,
+                                          cell_id_view);
+
+          for (unsigned int i = 0; i < dofs_per_cell; ++i)
+            local_to_global_host(cell_id, i) = lexicographic_dof_indices[i];
+
+          fe_values.reinit(*cell);
+
+          // Quadrature points
+          if (update_flags & update_quadrature_points)
+            {
+              const std::vector<Point<dim>> &q_points_vec =
+                fe_values.get_quadrature_points();
+              for (unsigned int i = 0; i < q_points_per_cell; ++i)
+                q_points_host(cell_id, i) = q_points_vec[i];
+            }
 
-      // Jacobian determinants/quadrature weights
-      if (update_flags & update_JxW_values)
-        {
-          data->JxW[color] = JxW;
-        }
+          if (update_flags & update_JxW_values)
+            {
+              std::vector<double> JxW_values_double =
+                fe_values.get_JxW_values();
+              for (unsigned int i = 0; i < q_points_per_cell; ++i)
+                JxW_host(cell_id, i) =
+                  static_cast<Number>(JxW_values_double[i]);
+            }
 
-      // Inverse jacobians
-      if (update_flags & update_gradients)
-        {
-          data->inv_jacobian[color] = inv_jacobian;
+          if (update_flags & update_gradients)
+            {
+              const std::vector<DerivativeForm<1, dim, dim>> &inv_jacobians =
+                fe_values.get_inverse_jacobians();
+              for (unsigned int i = 0; i < q_points_per_cell; ++i)
+                for (unsigned int j = 0; j < dim; ++j)
+                  for (unsigned int k = 0; k < dim; ++k)
+                    inv_jacobian_host(cell_id, i, j, k) =
+                      inv_jacobians[i][j][k];
```

Comment:
Same here, and switch to more common loop variable names
```suggestion
              for (unsigned int i = 0; i < q_points_per_cell; ++i)
                for (unsigned int d = 0; d < dim; ++d)
                  for (unsigned int e = 0; e < dim; ++e)
                    inv_jacobian_host(cell_id, i, d, e) =
                      fe_values.inverse_jacobian(i)[j][k];
```